File: user-manager/user-manager-rest-api/src/config/AbstractConfigManager.ts
import { ConfigurationCastException } from './errors/ConfigurationCastException';
import { ConfigurationUndefinedException } from './errors/ConfigurationUndefinedException';

export abstract class AbstractConfigManager {
  /**
   * Reads a value from a configuration resource.
   * @param property the configuration property name.
   * @return the configuration value corresponding to the given property.
   */
  protected abstract readConfigValue(property: string): string | undefined;

  /**
   * Returns the value of the configuration variable specified by `property`, or the value specified by
   * `defaultProperty` if `property` isn't set.
   * If the configuration variable identified by `property` isn't set and `defaultProperty` is null,
   * an instance of `ConfigurationUndefinedException` is thrown.
   *
   * @param property property the configuration property name.
   * @param defaultProperty the default value returned if the configuration identified by `property` isn't set.
   */
  getStringProperty(property: string, defaultProperty?: string): string {
    return this.getConfigValue(property, defaultProperty, AbstractConfigManager.stringToStringMapper, 'string');
  }

  /**
   * Returns the value of the configuration variable specified by `property` casted to boolean, or the
   * value specified by `defaultProperty` if `property` isn't set.
   * If the configuration variable identified by `property` isn't set and `defaultProperty` is null,
   * an instance of `ConfigurationUndefinedException` is thrown.
   * If the value of <code>property</code> cannot be casted to boolean, an instance of
   * <code>ConfigurationCastException</code> is thrown.
   *
   * @param property property the configuration property name.
   * @param defaultProperty the default value returned if the configuration identified by `property` isn't set.
   */
  getBooleanProperty(property: string, defaultProperty?: boolean): boolean {
    return this.getConfigValue(property, defaultProperty, AbstractConfigManager.stringToBooleanMapper, 'boolean');
  }

  /**
   * Returns the value of the configuration variable specified by `property` casted to integer, or the
   * value specified by `defaultProperty` if `property` isn't set.
   * If the configuration variable identified by `property` isn't set and `defaultProperty` is null,
   * an instance of `ConfigurationUndefinedException` is thrown.
   * If the value of <code>property</code> cannot be casted to integer, an instance of
   * <code>ConfigurationCastException</code> is thrown.
   *
   * @param property property the configuration property name.
   * @param defaultProperty the default value returned if the configuration identified by `property` isn't set.
   */
  getIntProperty(property: string, defaultProperty?: number): number {
    return this.getConfigValue(property, defaultProperty, AbstractConfigManager.stringToIntegerMapper, 'number');
  }

  /**
   * If both the environment variable identified by `property` and `defaultProperty` do not exist,
   * it throws `ConfigurationError`.
   * @param property
   * @param defaultProperty
   * @param mapper
   */
  private getConfigValue<T>(
    property: string,
    defaultProperty: T,
    mapper: (value: string) => NonNullable<T>,
    type: string,
  ): NonNullable<T> {
    const propertyValue = this.readConfigValue(property);
    if (propertyValue == null) {
      if (defaultProperty == null) {
        throw new ConfigurationUndefinedException(property);
      } else {
        // if the variable isn't set but at least there's a default value
        return defaultProperty!;
      }
    } else {
      // if the variable is set, try to cast it
      try {
        return mapper(propertyValue);
      } catch (_) {
        throw new ConfigurationCastException(property, propertyValue, type);
      }
    }
  }

  /**
   * stringToStringMapper returns an identity.
   */
  private static stringToStringMapper(value: string): string {
    return value;
  }

  /**
   * stringToBooleanMapper attempts to casts a String value to Boolean.
   */
  private static stringToBooleanMapper(value: string): boolean {
    const valueLowerCase = value.toLowerCase();
    if (valueLowerCase === 'true') {
      return true;
    } else if (valueLowerCase === 'false') {
      return false;
    } else {
      throw new TypeError();
    }
  }

  /**
   * stringToIntegerMapper attempts to cast a String value to Integer.
   */
  private static stringToIntegerMapper(value: string): number {
    const result = Number.parseInt(value, 10);
    if (Number.isNaN(result)) {
      throw new TypeError();
    }
    return result;
  }
}

File: butterfly/config/src/main/java/it/unipd/dstack/butterfly/config/AbstractConfigManager.java
package it.unipd.dstack.butterfly.config;

import it.unipd.dstack.butterfly.config.exception.ConfigurationCastException;
import it.unipd.dstack.butterfly.config.exception.ConfigurationException;
import it.unipd.dstack.butterfly.config.exception.ConfigurationUndefinedException;

import java.util.Optional;
import java.util.function.Function;

/**
 * AbstractConfigManager is an abstract class that defines the contract for reading configuration properties,
 * setting default values and explicitly cast them to the following types:
 * - String;
 * - Integer;
 * - Boolean;
 * It implements the Template Pattern via the readConfigValue method.
 * In case of errors, an instance of <code>ConfigurationException</code> is thrown.
 */
public abstract class AbstractConfigManager {
    /**
     * Reads a value from a configuration resource.
     * @param property
     * @return the configuration value corresponding to the given property.
     */
    protected abstract String readConfigValue(String property);

    /**
     * Returns the value of the configuration variable specified by <code>property</code>.
     * If the configuration variable identified by <code>property</code> isn't set, an instance of
     * <code>ConfigurationUndefinedException</code> is thrown.
     *
     * @param property the configuration property name
     * @return the value of the specified configuration variable.
     */
    public final String getStringProperty(String property) throws ConfigurationUndefinedException {
        return this.getStringProperty(property, null);
    }

    /**
     * Returns the value of the configuration variable specified by <code>property</code>, or the value specified by
     * <code>defaultProperty</code> if <code>property</code> isn't set.
     * If the configuration variable identified by <code>property</code> isn't set and defaultProperty is null,
     * an instance of <code>ConfigurationUndefinedException</code> is thrown.
     *
     * @param property
     * @param defaultProperty
     * @return the value of the specified configuration variable, or defaultProperty if it isn't set.
     */
    public final String getStringProperty(String property, String defaultProperty) throws ConfigurationUndefinedException {
        return this
                .getConfigValue(property, defaultProperty, AbstractConfigManager.stringToStringMapper, String.class);
    }

    /**
     * Returns the value of the configuration variable specified by <code>property</code> casted to Boolean.
     * If the configuration variable identified by <code>property</code> isn't set, an instance of
     * <code>ConfigurationUndefinedException</code> is thrown.
     * If the value of <code>property</code> cannot be casted to Boolean, an instance of
     * <code>ConfigurationCastException</code> is thrown.
     *
     * @param property the configuration property name
     * @return the value of the specified configuration variable.
     */
    public final Boolean getBooleanProperty(String property) throws ConfigurationException {
        return this.getBooleanProperty(property, null);
    }

    /**
     * Returns the value of the configuration variable specified by <code>property</code> casted to Boolean, or the
     * value specified by <code>defaultProperty</code> if <code>property</code> isn't set.
     * If the configuration variable identified by <code>property</code> isn't set and defaultProperty is null,
     * an instance of <code>ConfigurationUndefinedException</code> is thrown.
     * If the value of <code>property</code> cannot be casted to Boolean, an instance of
     * <code>ConfigurationCastException</code> is thrown.
     *
     * @param property
     * @param defaultProperty
     * @return the value of the specified configuration variable, or defaultProperty if it isn't set.
     */
    public final Boolean getBooleanProperty(String property, Boolean defaultProperty) throws ConfigurationException {
        return this
                .getConfigValue(property, defaultProperty, AbstractConfigManager.stringToBooleanMapper, Boolean.class);
    }

    /**
     * Returns the value of the configuration variable specified by <code>property</code> casted to Integer.
     * If the configuration variable identified by <code>property</code> isn't set, an instance of
     * <code>ConfigurationUndefinedException</code> is thrown.
     * If the value of <code>property</code> cannot be casted to Integer, an instance of
     * <code>ConfigurationCastException</code> is thrown.
     *
     * @param property the configuration property name
     * @return the value of the specified configuration variable.
     */
    public final Integer getIntProperty(String property) throws ConfigurationException {
        return this.getIntProperty(property, null);
    }

    /**
     * Returns the value of the configuration variable specified by <code>property</code> casted to Integer, or the
     * value specified by <code>defaultProperty</code> if <code>property</code> isn't set.
     * If the configuration variable identified by <code>property</code> isn't set and defaultProperty is null,
     * an instance of <code>ConfigurationUndefinedException</code> is thrown.
     * If the value of <code>property</code> cannot be casted to Integer, an instance of
     * <code>ConfigurationCastException</code> is thrown.
     *
     * @param property
     * @param defaultProperty
     * @return the value of the specified configuration variable, or defaultProperty if it isn't set.
     */
    public final Integer getIntProperty(String property, Integer defaultProperty) throws ConfigurationException {
        return this
                .getConfigValue(property, defaultProperty, AbstractConfigManager.stringToIntegerMapper, Integer.class);
    }

    /**
     * Retrieves the value of the configuration identified by <code>property</code> wrapped in an Optional object.
     *
     * @param property the configuration property name
     * @return the optional configuration value
     */
    private Optional<String> getOptionalConfigValue(String property) {
        return Optional.ofNullable(this.readConfigValue(property));
    }

    /**
     * Returns the value of the configuration variable specified by <code>property</code> casted via the
     * <code>mapper</code> function, or the value specified by <code>defaultProperty</code> if <code>property</code>
     * isn't set.
     *
     * @param property        the name of the configuration variable to read.
     * @param defaultProperty the value to be returned if <code>property</code> is not defined.
     * @param mapper          function used to cast the value of the configuration property to the desired type
     *                        <code>T</code>.
     * @param type            class name of the destination type, only used in case ConfigurationCastException is
     *                        thrown.
     * @return the value of the variable casted to type <code>T</code>, or <code>defaultProperty</code> if the
     * variable is not defined in the system environment
     */
    private <T> T getConfigValue(String property, T defaultProperty, Function<String, T> mapper, Class<?> type) {
        Optional<String> propertyValue = this.getOptionalConfigValue(property);
        if (propertyValue.isEmpty()) {
            if (defaultProperty == null) {
                throw new ConfigurationUndefinedException(property);
            } else {
                // if the variable isn't set but at least there's a default value
                return defaultProperty;
            }
        } else {
            // if the variable is set, try to cast it
            try {
                return mapper.apply(propertyValue.get());
            } catch (ClassCastException | NumberFormatException e) {
                throw new ConfigurationCastException(property, propertyValue.get(), type);
            }
        }
    }

    /**
     * stringToStringMapper returns an identity.
     */
    private static Function<String, String> stringToStringMapper = Function.identity();

    /**
     * stringToBooleanMapper attempts to casts a String value to Boolean.
     */
    private static Function<String, Boolean> stringToBooleanMapper = (String value) -> Boolean.valueOf(value);

    /**
     * stringToIntegerMapper attempts to cast a String value to Integer.
     */
    private static Function<String, Integer> stringToIntegerMapper = (String value) -> Integer.valueOf(value);
}
